python/tester/upload.py

```diff
@@ -84,34 +153,222 @@ def main():
         "./prompt-proto-upload.json"
     )
     storage_client = storage.Client(PROJECT_ID, credentials=credentials)
-    bucket = storage_client.bucket("rubin-prompt-proto-main")
+    dest_bucket = storage_client.bucket("rubin-prompt-proto-main")
     batch_settings = pubsub_v1.types.BatchSettings(
         max_messages=INSTRUMENTS[instrument].n_detectors,
     )
     publisher = pubsub_v1.PublisherClient(credentials=credentials,
                                           batch_settings=batch_settings)
 
+    last_group = get_last_group(storage_client, instrument, date)
+    _log.info(f"Last group {last_group}")
+
+    src_bucket = storage_client.bucket("rubin-prompt-proto-unobserved")
+    raw_pool = get_samples(src_bucket, instrument)
+
+    if raw_pool:
+        _log.info(f"Observing real raw files from {instrument}.")
+        for i, true_group in enumerate(itertools.islice(itertools.cycle(raw_pool), n_groups)):
+            group = last_group + i + 1
+            _log.debug(f"Processing group {group} from unobserved {true_group}...")
+            # snap_dict maps snap_id to {visit: blob}
+            snap_dict = {}
+            for snap_id, old_visits in raw_pool[true_group].items():
+                snap_dict[snap_id] = {splice_group(true_visit, group): blob
+                                      for true_visit, blob in old_visits.items()}
+            visit_infos = {info for det_dict in snap_dict.values() for info in det_dict}
```

Comment:
The above `for` loop and dict comps seem difficult to read for me, but I'm not sure that there's anything to be done about it. Maybe at least a comment explaining what this block is supposed to do?